README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mmibain

<!-- badges: start -->
<!-- badges: end -->

The Mighty Metrika Interface to BAIN (‘mmibain’) R package provides
Shiny apps to explore basic functionality of the
[‘bain’](https://informative-hypotheses.sites.uu.nl/software/bain/)
package for BAyesian INformative Hypotheses Evaluation.

## Installation

You can install the released version of ‘mmibain’ from
[CRAN](https://CRAN.R-project.org):

``` r
install.packages("mmibain")
```

You can install the development version of ‘mmibain’ from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mightymetrika/mmibain")
```

## Play RepliCrisis

‘RepliCrisis’ is a Shiny app game that simulates evalutating replication
studies based on the framework presented in [Hoijtink, Mulder, van Lissa
& Gu (2019)](https://doi.org/10.1037/met0000201). Follow these steps to
play:

- Set your sample size (for groups within study), difficulty, alpha
  level, and seed for reproducibility.
- Define thresholds for the Bayes Factor and Posterior Model Probability
  to assess evidence in favor of the original study.
- Conduct the original study to generate data and form a hypothesis.
- Show diagnostics and descriptives to understand statistical results
  and hypotheses.
- Conduct a replication study, using swap controls to match the original
  study’s results.
- Run replication analysis to evaluate the results against the original
  hypothesis.
- Start a new game by conducting a new original study.

To play, load ‘mmibain’ and call the RepliCrisis() function:

``` r
library(mmibain)
RepliCrisis()
```

## mmibain Shiny App

The package also includes a Shiny app for running basic bain::bain()
models:

- Upload your data in CSV format.
- Choose your modeling engine (lm, t_test, lavaan).
- Input your model and any additional arguments.
- Fit the model and input hypotheses for evaluation.
- Adjust settings such as the fraction parameter, standardized
  regression coefficients, and confidence intervals.
- Set a seed for reproducible results.
- Run the Bayesian Informative Hypotheses Evaluation.

Launch the app with:

``` r
mmibain()
```

# References

Hoijtink, H., Mulder, J., van Lissa, C., & Gu, X. (2019). A tutorial on
testing hypotheses using the Bayes factor. Psychological methods, 24(5),
539–556. <https://doi.org/10.1037/met0000201>


NEWS.md
# mmibain (development version)

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(mmibain)

test_check("mmibain")


tests/testthat/test-mmib_model.R
test_that("mmib_model works with the `lm` engine", {
  # Fit linear model
  mod <- mmib_model(mpg ~ wt + qsec, data = mtcars, engine = "lm")
  mod_comp <- stats::lm(mpg ~ wt + qsec, data = mtcars)

  expect_equal(sum(mod$coefficients == mod_comp$coefficients), 3)
  expect_equal(class(mod), class(mod_comp))

  expect_error(mmib_model(mpg ~ wt + qsec, data = mtcars))
  expect_error(mmib_model(mpg ~ wt + qsec, column_names = "mpg", data = mtcars,
                          engine = "lm"), "Only one of formula, column_names, or model should be provided.")
})

test_that("mmib_model works with the `t_test` engine", {
  # Fit t.test with formula
  expect_error(mmib_model(formula = vs ~ am, data = mtcars, engine = "t_test"))

  # Fit t.test with columns
  mod <- mmib_model(column_names = c("vs", "am"), data = mtcars, engine = "t_test")
  mod_comp <- stats::t.test(mtcars$vs, mtcars$am)

  expect_true(mod$statistic == mod_comp$statistic)
  expect_equal(class(mod), class(mod_comp))

  rm(mod, mod_comp)

  # Fit t.test with reverse columns
  mod <- mmib_model(column_names = c("vs", "am"), data = mtcars, engine = "t_test")
  mod_comp <- stats::t.test(mtcars$am, mtcars$vs)

  expect_true(mod$statistic == -mod_comp$statistic)

})

test_that("mmib_model works with the `lavaan` engine", {

  #  Specify model
  mod <- '
    A =~ mpg + wt + drat + cyl
    B =~ disp + hp + cyl
  '

  # Fit models
  suppressWarnings(fit <- mmib_model(model = mod, data = mtcars, engine = "lavaan",
                                     std.lv = TRUE))
  suppressWarnings(fit_comp <- lavaan::sem(mod, data = mtcars, std.lv = TRUE))


  expect_true(fit@test$standard$stat == fit_comp@test$standard$stat)
})


tests/testthat/test-rc_game_mgmt.R
test_that("Orginial Study Workflow", {

  set.seed(100)

  # Run basic work flow
  result <- deal_cards_to_rc_grid(n = 3)
  study_data <- generate_study_data(result, sample_size = 100)
  study_results <- process_original_study(study_data)

  # Test expectations
  expect_equal(length(result), 6)
  expect_s3_class(study_data, "data.frame")
  expect_s3_class(study_results$fit, "aov")
  expect_s3_class(study_results$fit, "lm")
})

test_that("Game Work Flow",{

  set.seed(250)

  # Run basic work flow

  ## Original study
  os_deck <- deal_cards_to_rc_grid(n = 3)
  original_study_data <- generate_study_data(os_deck, sample_size = 100)
  original_study_results <- process_original_study(original_study_data)

  ## Replication study
  rs_deck <- deal_cards_to_rc_grid(n = 3)
  replication_data <- generate_study_data(rs_deck, sample_size = 100)
  replication_results <- process_replication_study(replication_data,
                                                   original_study_results)


  # Test expectations
  expect_s3_class(replication_results$bain_results, "bain")

})


test_that("Interpret Replications Results",{

  set.seed(1220)

  # Run basic work flow

  ## Original study
  os_deck <- deal_cards_to_rc_grid(n = 3)
  original_study_data <- generate_study_data(os_deck, sample_size = 100)
  original_study_results <- process_original_study(original_study_data)

  ## Replication study
  rs_deck <- deal_cards_to_rc_grid(n = 3)
  replication_data <- generate_study_data(rs_deck, sample_size = 100)
  replication_results <- process_replication_study(replication_data,
                                                   original_study_results)

  interp <- interpret_replication_results(replication_results)


  # Test expectations
  expect_equal(length(interp), 3)
})


test_that("Card Swap",{

  set.seed(492)

  # Replication study

  ## Swap columns
  rs_deck <- deal_cards_to_rc_grid(n = 3)
  s1 <- swapper(cards_matrix = rs_deck, c(1,2))
  expect_equal(rs_deck[1,1][[1]]$value, s1[1,2][[1]]$value)
  expect_equal(rs_deck[1,2][[1]]$value, s1[1,1][[1]]$value)
  expect_s3_class(s1, "swapper")

  ## Swap within columns
  s2 <- swapper(cards_matrix = s1, swap_in_col = 3)
  expect_equal(s1[1,3][[1]]$value, s2[2,3][[1]]$value)
  expect_equal(s1[2,3][[1]]$value, s2[1,3][[1]]$value)
  expect_equal(attr(s2, "swap_in_col_hist"), 1)
  expect_s3_class(s2, "swapper")

  ## Swap columns 2 and 3 within row 1
  s3 <- swapper(cards_matrix = s2, swap_in_row = c(1, 2, 3))
  expect_equal(s2[1,2][[1]]$value, s3[1,3][[1]]$value)
  expect_equal(s2[1,3][[1]]$value, s3[1,2][[1]]$value)
  expect_equal(attr(s3, "swap_in_col_hist"), 1)
  expect_equal(attr(s3, "swap_in_row1_hist"), 1)
  expect_equal(attr(s3, "swap_in_row2_hist"), 0)

  ## Swap columns 1 and 3 within row 2
  s4 <- swapper(cards_matrix = s3, swap_in_row = c(2, 1, 3))
  expect_equal(s3[2,1][[1]]$value, s4[2,3][[1]]$value)
  expect_equal(s3[2,3][[1]]$value, s4[2,1][[1]]$value)
  expect_equal(attr(s4, "swap_in_col_hist"), 1)
  expect_equal(attr(s4, "swap_in_row1_hist"), 1)
  expect_equal(attr(s4, "swap_in_row2_hist"), 1)

  ## Try a second swap within columns
  expect_error(swapper(cards_matrix = s4, swap_in_col = 2), "You can't swap within columns more than once.")

  ## Try a second swap within row1
  expect_error(swapper(cards_matrix = s4, swap_in_row = c(1, 1, 3)), "You can't swap within row 1 more than once.")

  ## Try a second swap within row2
  expect_error(swapper(cards_matrix = s4, swap_in_row = c(2, 2, 3)), "You can't swap within row 2 more than once.")


})


test_that("i_deck Game Workflow",{

  set.seed(250)

  # Run basic work flow

  ## Original study
  os_deck <- deal_cards_to_rc_grid(n = 3)
  original_study_data <- generate_study_data(os_deck, sample_size = 100)
  original_study_results <- process_original_study(original_study_data)

  ## Replication study
  rs_deck <- deal_cards_to_rc_grid(n = 3)
  replication_data <- generate_study_data(rs_deck, sample_size = 100)
  replication_results <- process_replication_study(replication_data,
                                                   original_study_results)

  # Test expectations
  expect_s3_class(replication_results$bain_results, "bain")

})

# test_that("Random Hands Game Flow",{
#
#   # Create function to test random hands
#   find_errors <- function(num_iterations, n = 3, sample_size = 10) {
#     error_seeds <- numeric(0)
#
#     for (i in 1:num_iterations) {
#       seed <- sample.int(10000, 1)  # Randomly selecting a seed value up to 10,000
#       set.seed(seed)
#
#       tryCatch({
#         ## Original study
#         os_deck <- deal_cards_to_rc_grid(n = n)
#         original_study_data <- generate_study_data(os_deck, sample_size = sample_size)
#         original_study_results <- process_original_study(original_study_data)
#
#         ## Replication study
#         rs_deck <- deal_cards_to_rc_grid(n = n)
#         replication_data <- generate_study_data(rs_deck, sample_size = sample_size)
#         replication_results <- process_replication_study(replication_data,
#                                                          original_study_results)
#       },
#       error = function(e) {
#         message("Error encountered with seed: ", seed)
#         error_seeds <- c(error_seeds, seed)
#       })
#     }
#
#     return(error_seeds)
#   }
#
#   # Call the function
#   num_iterations <- 5
#
#   # Test on n = 3, sample_size = 10
#   error_seeds_3_10 <- find_errors(num_iterations)
#   expect_equal(length(error_seeds_3_10), 0)
#
#   # Test on n = 13, sample_size = 15
#   error_seeds_13_15 <- find_errors(num_iterations, n = 13, sample_size = 15)
#   expect_equal(length(error_seeds_13_15), 0)
#
# })


R/helpers.R
#' Convert a Character String to a List
#'
#' This internal function converts a character string representing R arguments
#' into a list of arguments. It is primarily used to facilitate the passing of
#' additional arguments from the Shiny UI to internal functions within the app.
#'
#' @param arg_str A character string representing R arguments.
#'
#' @return A list containing the arguments represented by \code{arg_str}. If
#'   \code{arg_str} is not a valid representation of R arguments, the function
#'   will throw an error.
#'
#' @keywords internal
str2list <- function(arg_str) {
  # Evaluate the string in a new environment to convert it to a list
  eval(parse(text = paste0("list(", arg_str, ")")), envir = new.env())
}

#' Generate Descriptive Statistics for Study Groups
#'
#' @description
#' This internal helper function generates descriptive statistics for each group
#' within the original study data. It calculates the sample size, mean, standard
#' deviation, median, median absolute deviation, minimum, maximum, skewness, and
#' kurtosis for the values associated with each group.
#'
#' @param df A data frame containing the study data.
#' @param group_var The name of the column in `df` that contains the group labels.
#'
#' @details
#' The function iterates over each unique group found in the `group_var` column
#' of the data frame `df`. For each group, it computes the descriptive statistics
#' using base R functions and functions from the `psych` and `e1071` packages for
#' skewness and kurtosis, respectively.
#'
#' The results are returned as a data frame where each row corresponds to a group
#' and each column to a descriptive statistic.
#'
#' @note
#' As this is an internal function intended for use within process_original_study
#' function of the package.
#'
#' @keywords internal
generate_descriptives <- function(df, group_var) {
  # List of unique groups
  groups <- unique(df[[group_var]])

  # Function to calculate descriptives for each group
  calc_descriptives <- function(group_data) {
    group_ColVals <- group_data$ColVals

    list(
      n = length(group_ColVals),
      mean = mean(group_ColVals, na.rm = TRUE),
      sd = stats::sd(group_ColVals, na.rm = TRUE),
      median = stats::median(group_ColVals, na.rm = TRUE),
      mad = stats::mad(group_ColVals, na.rm = TRUE),
      min = min(group_ColVals, na.rm = TRUE),
      max = max(group_ColVals, na.rm = TRUE),
      skew = psych::skew(group_ColVals, na.rm = TRUE),
      kurtosis = e1071::kurtosis(group_ColVals, na.rm = TRUE)
    )
  }

  # Apply the function to each group
  descriptives_list <- lapply(groups, function(group) {
    group_data <- df[df[[group_var]] == group, ]
    c(list(Group = group), calc_descriptives(group_data))
  })

  # Convert list to data frame
  descriptives_df <- do.call(rbind.data.frame, descriptives_list)

  return(descriptives_df)
}


R/mmib_model.R
#' Fit Statistical Models for MMI Bain Processing
#'
#' This function provides a unified interface to fit different statistical models
#' supported by the 'bain' package.
#'
#' @param formula A symbolic description of the model to be fit. Used specifically
#' for the `lm` engine. Default is NULL.
#' @param column_names A character vector of length 2, representing the column
#' names to be used for the `t_test` engine. Default
#' is NULL.
#' @param model A model specification (usually as a string) for the `lavaan` engine.
#' Default is NULL.
#' @param data A data frame containing the variables in the model.
#' @param engine A character string representing the statistical method to be used.
#' Currently supported methods are: `"lm"`, `"t_test"`, and `"lavaan"`.
#' @param ... Additional arguments to be passed to the underlying statistical
#' function (lm(), t.test(), or lavaan::sem()).
#'
#' @return Returns an object of the type associated with the engine selected (`lm`,
#' `htest`, or `lavaan` object).
#'
#' @details
#' The mmib_model() function provides a simple interface to fit various statistical
#' models, which can be subsequently processed by the bain::bain() function. It
#' ensures that only one of `formula`, `column_names`, or `model` is provided,
#' checks the validity of the provided data, and selects the appropriate statistical
#' method based on the `engine` parameter.
#'
#' @examples
#' data(mtcars)
#'
#' # Fit linear model
#' mod1 <- mmib_model(mpg ~ wt + qsec, data = mtcars, engine = "lm")
#'
#' @seealso \code{\link[bain]{bain}} for processing the models fit with `mmib_model`.
#'
#' @export
mmib_model <- function(formula = NULL, column_names = NULL, model = NULL, data,
                       engine = c("lm", "t_test", "lavaan"), ...) {

  # Check inputs
  if (sum(!is.null(formula), !is.null(column_names), !is.null(model)) != 1) {
    stop("Only one of formula, column_names, or model should be provided.")
  }

  if (!is.data.frame(data)) {
    stop("data is not a valid data frame")
  }

  # Fit the model based on the engine selected
  switch(engine,
         lm = {
           if (is.null(formula)) stop("Formula is required for lm engine.")
           model <- stats::lm(formula = formula, data, ...)
         },
         t_test = {
           if (!is.null(column_names) && length(column_names) == 2) {
             x <- data[[column_names[1]]]
             y <- data[[column_names[2]]]
             model <- stats::t.test(x, y, ...)
           } else {
             stop("For t_test provide two column names.")
           }
         },
         lavaan = {
           if (is.null(model)) stop("Model specification is required for lavaan engine.")
           model <- lavaan::sem(model = model, data = data, ...)
         },
         stop(paste("Invalid engine:", engine))
  )

  return(model)
}


R/RepliCrisis.R
#' RepliCrisis Shiny App
#'
#' The RepliCrisis Shiny app is an interactive game based on the process for
#' evaluating replication studies presented in Hoijtink et. al. (2009). Users can
#' conduct original studies, process the results, and attempt to replicate the
#' studies while adjusting parameters to see how changes affect the outcome.
#'
#'
#' @return An interactive Shiny app.
#'
#' @examples
#' if(interactive()){
#' RepliCrisis()
#' }
#'
#' @export
#' @keywords shinyapp
#'
#' @references
#' Hoijtink, H., Mulder, J., van Lissa, C., & Gu, X. (2019). A tutorial on
#' testing hypotheses using the Bayes factor. Psychological methods, 24(5),
#' 539–556. https://doi.org/10.1037/met0000201
RepliCrisis <- function(){
  ui <- shiny::fluidPage(
    theme = shinythemes::shinytheme("paper"),
    shiny::titlePanel("RepliCrisis"),

    shiny::sidebarLayout(
      shiny::sidebarPanel(
        # Game setup
        shiny::numericInput("sample_size", "Select Sample Size", value = 10, min = 3, max = 500),
        shiny::numericInput("difficulty", "Select Difficulty", value = 3, min = 3, max = 13),
        shiny::numericInput("alpha_level", "Select Alpha Level", value = 0.05, min = 0.01, max = 0.10, step = 0.01),
        shiny::numericInput("seed_value", "Set seed (optional)", value = NA, min = 1, max = .Machine$integer.max),
        shiny::numericInput("bf_threshold", "Bayes Factor Threshold", value = 10),
        shiny::numericInput("pmpb_threshold", "Posterior Model Probability Threshold", value = 0.95),

        # Original study
        shiny::fluidRow(
          shiny::column(12, shiny::actionButton("original_study", "Conduct Original Study")),
          shiny::column(12, shiny::actionButton("show_diagnostics", "Show Diagnostics")),
          shiny::column(12, shiny::actionButton("show_descriptives", "Show Descriptives")),

          # Replication study
          shiny::column(12, shiny::actionButton("replication_study", "Conduct Replication Study"))
        ),
        shiny::actionButton("run_replication", "Run Replication Analysis")
      ),

      shiny::mainPanel(
        shiny::uiOutput("original_study_title"),
        shiny::uiOutput("model_summary_title"),
        shiny::uiOutput("original_hypothesis_title"),
        shiny::uiOutput("diagnostics_title"),
        shiny::uiOutput("descriptive_stats_title"),
        shiny::uiOutput("replication_study_title"),
        shiny::uiOutput("swap_cards_title"),
        shiny::uiOutput("results_title"),
      )
    )
  )

  server <- function(input, output, session) {

    # Set up reactive values
    original_conducted <- shiny::reactiveVal(FALSE)
    study_results <- shiny::reactiveVal(NULL)
    diagnostics_requested <- shiny::reactiveVal(FALSE)
    descriptives_requested <- shiny::reactiveVal(FALSE)
    replication_cards <- shiny::reactiveVal(NULL)
    replication_conducted <- shiny::reactiveVal(FALSE)
    replication_results <- shiny::reactiveVal(NULL)

    shiny::observeEvent(input$original_study, {

      # Reset reactive values
      original_conducted(FALSE)
      study_results(NULL)
      diagnostics_requested(FALSE)
      descriptives_requested(FALSE)
      replication_cards(NULL)
      replication_conducted(FALSE)
      replication_results(NULL)

      # If seed value is provided, set it
      if (!is.na(input$seed_value)) {
        set.seed(input$seed_value)
      }

      # Deal cards
      card_grid <- deal_cards_to_rc_grid(n = input$difficulty)

      # Rendering the UI for the card grid
      output$card_display <- shiny::renderUI({
        render_card_grid(card_grid)
        })

      # Generate study data
      study_data <- generate_study_data(card_grid, sample_size = input$sample_size)

      # Process original study
      study_results <- process_original_study(study_data, alpha = input$alpha_level)

      # Update study_results to hold the generated data
      study_results(study_results)

      # Display fit summary and hypothesis
      output$fit_summary <- shiny::renderTable(broom::tidy(study_results$fit))
      output$pair_t <- shiny::renderTable(broom::tidy(study_results$pairwise_t))
      output$hypothesis <- shiny::renderText(study_results$hypothesis)

      # Update study conducted reactiveVal
      original_conducted(TRUE)
    })

    # Conditional UI for fit summary
    output$fit_summary_ui <- shiny::renderUI({
      if(input$original_study) {
        shiny::tableOutput("fit_summary")
      }
    })

    # Conditional UI for pairwise t-test results
    output$pair_t_ui <- shiny::renderUI({
      if(input$original_study) {
        shiny::tableOutput("pair_t")
      }
    })

    # Conditional UI for hypothesis
    output$hypothesis_ui <- shiny::renderUI({
      if(input$original_study) {
        shiny::textOutput("hypothesis")
      }
    })

    shiny::observeEvent(input$show_diagnostics, {
      # Update reactiveVal
      diagnostics_requested(TRUE)

      # Display diagnostics for the original study: fit plot, Shapiro test, and Levene test

      # Extract study_results
      results <- study_results()

      output$fit_plot <- shiny::renderPlot({

        # Setup on.exit to reset graphical parameters
        oldpar <- graphics::par(no.readonly = TRUE)
        on.exit(graphics::par(oldpar))

        # Set up a 2x2 grid for 4 plots
        graphics::par(mfrow = c(2, 2))

        # Plot 1: Residuals vs Fitted
        plot(results$fit, which = 1)

        # Plot 2: Normal Q-Q
        plot(results$fit, which = 2)

        # Plot 3: Scale-Location (Standardized residuals vs Fitted values)
        plot(results$fit, which = 3)

        # Plot 4: Cook's distance
        plot(results$fit, which = 4)
      })

      # Get diagnostic statistical tests
      output$shapiro_test <- shiny::renderTable(broom::tidy(results$shapiro_test))
      lev_tst <- broom::tidy(results$levene_test)
      lev_tst$method <- "Levene Test"
      output$levene_test <- shiny::renderTable(lev_tst)

    })

    # Conditional UI for diagnostics
    output$fit_plot_ui <- shiny::renderUI({
      if(input$show_diagnostics) {
        shiny::plotOutput("fit_plot")
      }
    })

    output$shapiro_test_ui <- shiny::renderUI({
      if(input$show_diagnostics) {
        shiny::tableOutput("shapiro_test")
      }
    })

    output$levene_test_ui <- shiny::renderUI({
      if(input$show_diagnostics) {
        shiny::tableOutput("levene_test")
      }
    })

    shiny::observeEvent(input$show_descriptives, {

      #Update reactiveVal
      descriptives_requested(TRUE)

      # Display descriptive statistics for the original study

      # Extract study_results
      results <- study_results()

      output$descriptives <- shiny::renderTable(results$descriptives)
    })

    output$descriptives_ui <- shiny::renderUI({
      if(input$show_descriptives) {
        shiny::tableOutput("descriptives")
      }
    })

    shiny::observeEvent(input$replication_study, {
      # Deal cards for replication study
      card_grid_replication <- deal_cards_to_rc_grid(n = input$difficulty)

      # Update the reactive value
      if (!replication_conducted()){
        replication_cards(card_grid_replication)
      }


      # Rendering the UI for the card grid for replication study
      output$rep_card_display <- shiny::renderUI({
        render_card_grid(replication_cards())
      })

      # Set the reactive value to TRUE since the replication study is now conducted
      replication_conducted(TRUE)
    })

    shiny::observeEvent(input$swap_cols, {
      # Extract the replication card grid from the reactive value
      rep_cards <- replication_cards()

      # Check for NULL values in user input and exit early if found
      if (is.null(input$swap_col1) || is.null(input$swap_col2)) return(NULL)

      # Swap the columns using the swapper function
      tryCatch({
        new_card_grid <- swapper(rep_cards, swap_cols = c(input$swap_col1, input$swap_col2))

        # Update the reactive value to hold the new card grid
        replication_cards(new_card_grid)

        # Rerender the UI for the card grid
        output$rep_card_display <- shiny::renderUI({
          render_card_grid(new_card_grid)
        })
      }, error = function(e) {
        # Handle the error by displaying a message
        shiny::showNotification(paste("An error occurred:", e$message), type = "error")
      })
    })

    shiny::observeEvent(input$swap_inside_col, {
      # Extract the replication card grid from the reactive value
      rep_cards <- replication_cards()

      # Check for NULL values in user input and exit early if found
      if (is.null(input$swap_in_col)) return(NULL)

      # Swap within the column using the swapper function
      tryCatch({
        new_card_grid <- swapper(rep_cards, swap_in_col = input$swap_in_col)

        # Update the reactive value to hold the new card grid
        replication_cards(new_card_grid)

        # Rerender the UI for the card grid
        output$rep_card_display <- shiny::renderUI({
          render_card_grid(new_card_grid)
        })
      }, error = function(e) {
        # Handle the error by displaying a message
        shiny::showNotification(paste("An error occurred:", e$message), type = "error")
      })
    })

    shiny::observeEvent(input$swap_inside_row, {
      # Extract the replication card grid from the reactive value
      rep_cards <- replication_cards()

      # Check for NULL values in user input and exit early if found
      if (is.null(input$swap_in_row_num) || is.null(input$swap_in_row_col1) || is.null(input$swap_in_row_col2)) return(NULL)

      # Swap within the row using the swapper function
      tryCatch({
        new_card_grid <- swapper(rep_cards, swap_in_row = c(input$swap_in_row_num, input$swap_in_row_col1, input$swap_in_row_col2))

        # Update the reactive value to hold the new card grid
        replication_cards(new_card_grid)

        # Rerender the UI for the card grid
        output$rep_card_display <- shiny::renderUI({
          render_card_grid(new_card_grid)
        })
      }, error = function(e) {
        # Handle the error by displaying a message
        shiny::showNotification(paste("An error occurred:", e$message), type = "error")
      })
    })
    # Render the swap controls only when the replication study has been conducted
    output$swap_controls_ui <- shiny::renderUI({
      if (replication_conducted()) {
        shiny::fluidRow(
          # First column: Swap Columns
          shiny::column(4,
                        shiny::tagList(
                          shiny::numericInput("swap_col1", "Swap Column 1", value = NULL),
                          shiny::numericInput("swap_col2", "Swap Column 2", value = NULL),
                          shiny::actionButton("swap_cols", "Swap Columns")
                        )
          ),
          # Second column: Swap Inside Column
          shiny::column(4,
                        shiny::tagList(
                          shiny::numericInput("swap_in_col", "Swap Inside Column", value = NULL),
                          shiny::actionButton("swap_inside_col", "Execute Inside Column Swap")
                        )
          ),
          # Third column: Swap Inside Row
          shiny::column(4,
                        shiny::tagList(
                          shiny::numericInput("swap_in_row_num", "Row Number", value = NULL),
                          shiny::numericInput("swap_in_row_col1", "Swap Row Column 1", value = NULL),
                          shiny::numericInput("swap_in_row_col2", "Swap Row Column 2", value = NULL),
                          shiny::actionButton("swap_inside_row", "Execute Inside Row Swap")
                        )
          )
        )
      }
    })



    shiny::observeEvent(input$run_replication, {
      # Extract the current card deck for replication study from the reactive value
      rep_cards <- replication_cards()

      # If there's no replication card deck, exit early
      if (is.null(rep_cards)) return(NULL)

      # Generate replication study data
      replication_data <- generate_study_data(rep_cards, sample_size = input$sample_size)

      # Process replication study
      replication_results <- process_replication_study(replication_data, study_results())

      # Update reactive value
      replication_results(replication_results)

      # Display the replication study results in the UI
      output$bain_results_summary <- shiny::renderPrint({
        summary(replication_results()$bain_results)
      })

      output$bain_results <- shiny::renderPrint({
        print(replication_results()$bain_results)
      })

      # Interpret the replication results using the thresholds provided by the user
      interpretation <- interpret_replication_results(
        replication_results(),
        bf_threshold = input$bf_threshold,
        pmpb_threshold = input$pmpb_threshold
      )

      # Display the interpretation and result message
      output$interpretation <- shiny::renderUI({
        if (interpretation$result == "win") {
          shiny::tags$div(
            shiny::tags$h1("You Win!", style = "color: green; font-size: 48px;"),
            shiny::tags$p(interpretation$interpretation)
          )
        } else {
          shiny::tags$div(
            shiny::tags$h1("You Lose!", style = "color: red; font-size: 48px;"),
            shiny::tags$p(interpretation$interpretation)
          )
        }
      })

      # Display the disclaimer
      output$disclaimer <- shiny::renderText({
        interpretation$disclaimer
      })
    })

    # Original Study Title
    output$original_study_title <- shiny::renderUI({
      if(input$original_study) {
        shiny::tagList(
          shiny::tags$h3("Original Study"),
          shiny::uiOutput("card_display")
        )
      }
    })

    # Model Summary Title
    output$model_summary_title <- shiny::renderUI({
      if(input$original_study & original_conducted()) {
        shiny::tagList(
          shiny::tags$h3("Model Summary"),
          shiny::uiOutput("fit_summary_ui"),
          shiny::uiOutput("pair_t_ui")
        )
      }
    })

    # Original Hypothesis Title
    output$original_hypothesis_title <- shiny::renderUI({
      if(input$original_study & original_conducted()) {
        shiny::tagList(
          shiny::tags$h3("Original Hypothesis"),
          shiny::uiOutput("hypothesis_ui")
        )
      }
    })

    # Diagnostics Title
    output$diagnostics_title <- shiny::renderUI({
      if(diagnostics_requested() & original_conducted()) {
        shiny::tagList(
          shiny::tags$h3("Diagnostics"),
          shiny::uiOutput("fit_plot_ui"),
          shiny::uiOutput("shapiro_test_ui"),
          shiny::uiOutput("levene_test_ui")
        )
      }
    })

    # Descriptive Statistics Title
    output$descriptive_stats_title <- shiny::renderUI({
      if(descriptives_requested() & original_conducted()) {
        shiny::tagList(
          shiny::tags$h3("Descriptive Statistics"),
          shiny::uiOutput("descriptives_ui")
        )
      }
    })

    # Replication Study Title
    output$replication_study_title <- shiny::renderUI({
      if(replication_conducted() & original_conducted()) {
        shiny::tagList(
          shiny::tags$h3("Replication Study"),
          shiny::uiOutput("rep_card_display")
        )
      }
    })

    # Swap Cards Title
    output$swap_cards_title <- shiny::renderUI({
      if(replication_conducted() & original_conducted()) {
        shiny::tagList(
          shiny::tags$h3("Swap Cards"),
          shiny::uiOutput("swap_controls_ui")
        )
      }
    })

    # Results Title
    output$results_title <- shiny::renderUI({
      if(replication_conducted() & original_conducted() & !is.null(replication_results())) {
        shiny::tagList(
          shiny::tags$h3("Results"),
          shiny::verbatimTextOutput("bain_results_summary"),
          shiny::verbatimTextOutput("bain_results"),
          shiny::uiOutput("interpretation"),
          shiny::textOutput("disclaimer")
        )
      }
    })

  }

  shiny::shinyApp(ui, server)
}
